core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannelManager.scala

```diff
@@ -287,15 +289,65 @@ class TransactionMarkerChannelManager(config: KafkaConfig,
     def appendCallback(error: Errors): Unit =
       error match {
         case Errors.NONE =>
-          trace(s"Completed transaction for ${txnLogAppend.transactionalId} with coordinator epoch ${txnLogAppend.coordinatorEpoch}, final state: state after commit: ${txnLogAppend.txnMetadata.state}")
+          trace(s"Completed transaction for ${txnLogAppend.transactionalId} with coordinator epoch ${txnLogAppend.coordinatorEpoch}, final state after commit: ${txnLogAppend.txnMetadata.state}")
 
         case Errors.NOT_COORDINATOR =>
           info(s"No longer the coordinator for transactionalId: ${txnLogAppend.transactionalId} while trying to append to transaction log, skip writing to transaction log")
 
         case Errors.COORDINATOR_NOT_AVAILABLE =>
-          warn(s"Failed updating transaction state for ${txnLogAppend.transactionalId} when appending to transaction log due to ${error.exceptionName}. retrying")
-          // enqueue for retry
-          txnLogAppendRetryQueue.add(txnLogAppend)
+          info(s"I am currently not available to append $txnLogAppend: possible causes include ${Errors.UNKNOWN_TOPIC_OR_PARTITION}, ${Errors.NOT_ENOUGH_REPLICAS}, ${Errors.NOT_ENOUGH_REPLICAS_AFTER_APPEND} and ${Errors.REQUEST_TIMED_OUT}")
+
+          val transactionalId = txnLogAppend.transactionalId
+          val coordinatorEpoch = txnLogAppend.coordinatorEpoch
+
+          txnStateManager.getTransactionState(transactionalId) match {
```

Comment:
Hmm.. The duplication of code similar to what we already have in `TransactionCoordinator` is unfortunate. I thought we might be able to refine the logic in `TransactionStateManager.appendTransactionToLog`. For example, we could add an additional argument which indicates that the append will be retried, in which case we won't remove the pending state upon error. Would that work?